Hey, all, the two fixes above have revealed a third bug in the code, in which an `UnknownProducerId` exception is thrown, but not properly treated as recoverable (by rebalancing).

Here's the exception I've seen:
```
[2019-11-28 17:02:47,894] ERROR [stream-soak-test-f5695908-c881-4c1f-814b-5db3e0ca09ff-StreamThread-3] stream-thread [stream-soak-test-f5695908-c881-4c1f-814b-5db3e0ca09ff-StreamThread-3] task [1_0] Failed to close producer due to the following error: (org.apache.kafka.streams.processor.internals.StreamTask)
 org.apache.kafka.streams.errors.StreamsException: task [1_0] Abort sending since an error caught with a previous record (timestamp 1574960233670) to topic stream-soak-test-KSTREAM-AGGREGATE-STATE-STORE-0000000019-changelog due to org.apache.kafka.common.errors.UnknownProducerIdException: This exception is raised by the broker if it could not locate the producer metadata associated with the producerId in question. This could happen if, for instance, the producer's records were deleted because their retention time had elapsed. Once the last records of the producerId are removed, the producer's metadata is removed from the broker, and future appends by the producer will return this exception.
	at org.apache.kafka.streams.processor.internals.RecordCollectorImpl.recordSendError(RecordCollectorImpl.java:143)
	at org.apache.kafka.streams.processor.internals.RecordCollectorImpl.access$500(RecordCollectorImpl.java:51)
	at org.apache.kafka.streams.processor.internals.RecordCollectorImpl$1.onCompletion(RecordCollectorImpl.java:202)
	at org.apache.kafka.clients.producer.KafkaProducer$InterceptorCallback.onCompletion(KafkaProducer.java:1348)
	at org.apache.kafka.clients.producer.internals.ProducerBatch.completeFutureAndFireCallbacks(ProducerBatch.java:230)
	at org.apache.kafka.clients.producer.internals.ProducerBatch.done(ProducerBatch.java:196)
	at org.apache.kafka.clients.producer.internals.Sender.failBatch(Sender.java:730)
	at org.apache.kafka.clients.producer.internals.Sender.failBatch(Sender.java:716)
	at org.apache.kafka.clients.producer.internals.Sender.completeBatch(Sender.java:674)
	at org.apache.kafka.clients.producer.internals.Sender.handleProduceResponse(Sender.java:596)
	at org.apache.kafka.clients.producer.internals.Sender.access$100(Sender.java:74)
	at org.apache.kafka.clients.producer.internals.Sender$1.onComplete(Sender.java:798)
	at org.apache.kafka.clients.ClientResponse.onComplete(ClientResponse.java:109)
	at org.apache.kafka.clients.NetworkClient.completeResponses(NetworkClient.java:562)
	at org.apache.kafka.clients.NetworkClient.poll(NetworkClient.java:554)
	at org.apache.kafka.clients.producer.internals.Sender.runOnce(Sender.java:335)
	at org.apache.kafka.clients.producer.internals.Sender.run(Sender.java:244)
	at java.lang.Thread.run(Thread.java:748)
 Caused by: org.apache.kafka.common.errors.UnknownProducerIdException: This exception is raised by the broker if it could not locate the producer metadata associated with the producerId in question. This could happen if, for instance, the producer's records were deleted because their retention time had elapsed. Once the last records of the producerId are removed, the producer's metadata is removed from the broker, and future appends by the producer will return this exception.
 [2019-11-28 17:02:48,124] INFO [stream-soak-test-f5695908-c881-4c1f-814b-5db3e0ca09ff-StreamThread-3] [Consumer clientId=stream-soak-test-f5695908-c881-4c1f-814b-5db3e0ca09ff-StreamThread-3-restore-consumer, groupId=null] Unsubscribed all topics or patterns and assigned partitions (org.apache.kafka.clients.consumer.KafkaConsumer)
 [2019-11-28 17:02:48,134] INFO [stream-soak-test-f5695908-c881-4c1f-814b-5db3e0ca09ff-StreamThread-3] stream-thread [stream-soak-test-f5695908-c881-4c1f-814b-5db3e0ca09ff-StreamThread-3] State transition from PENDING_SHUTDOWN to DEAD (org.apache.kafka.streams.processor.internals.StreamThread)
 [2019-11-28 17:02:48,134] INFO [stream-soak-test-f5695908-c881-4c1f-814b-5db3e0ca09ff-StreamThread-3] stream-thread [stream-soak-test-f5695908-c881-4c1f-814b-5db3e0ca09ff-StreamThread-3] Shutdown complete (org.apache.kafka.streams.processor.internals.StreamThread)
```

Therefore, I've introduced the latest commit to this PR and re-started my test.

--- a/streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java
+++ b/streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java
@@ -177,7 +177,6 @@ public abstract class AbstractTask implements Task {
             stateMgr.flush();
         } catch (final ProcessorStateException e) {
             if (e.getCause() instanceof RecoverableClientException) {
-                log.warn("Caught a recoverable Kafka exception while flushing state. Initiating a rebalance to attempt recovery.", e);
                 throw new TaskMigratedException(this, e);
             }
         }

--- a/streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java
+++ b/streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java
@@ -188,10 +188,11 @@ public class RocksDBStore implements KeyValueStore<Bytes, byte[]>, BulkLoadingSt
             throw new ProcessorStateException(fatal);
         }
 
-        maybeSetUpMetricsRecorder(context, configs);
-
         openRocksDB(dbOptions, columnFamilyOptions);
         open = true;
+
+        // Do this last because the prior operations could throw exceptions.
+        maybeSetUpMetricsRecorder(context, configs);
     }
 
     private void maybeSetUpMetricsRecorder(final ProcessorContext context, final Map<String, Object> configs) {

--- a/streams/src/test/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasksTest.java
+++ b/streams/src/test/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasksTest.java
@@ -37,6 +37,7 @@ import org.apache.kafka.test.MockSourceNode;
 import org.easymock.EasyMock;
 import org.junit.Before;
 import org.junit.Test;
+import org.junit.function.ThrowingRunnable;
 
 import java.util.ArrayList;
 import java.util.Collection;
@@ -44,8 +45,8 @@ import java.util.Collections;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Set;
-import org.junit.function.ThrowingRunnable;
 
+import static org.hamcrest.CoreMatchers.is;
 import static org.hamcrest.CoreMatchers.not;
 import static org.hamcrest.CoreMatchers.nullValue;
 import static org.hamcrest.MatcherAssert.assertThat;
@@ -235,12 +236,43 @@ public class AssignedStreamsTasksTest {
         t1.suspend();
         EasyMock.expectLastCall().andThrow(new TaskMigratedException());
         t1.close(false, true);
-        EasyMock.expectLastCall();
+        EasyMock.expectLastCall().andThrow(new RuntimeException("any exception"));
         EasyMock.replay(t1);
 
         assertThat(suspendTask(), nullValue());
         assertTrue(assignedTasks.runningTaskIds().isEmpty());
         EasyMock.verify(t1);
+    }
+
+    @Test
+    public void shouldCloseUncleanAndThenRethrowOnShutdownIfRuntimeException() {
+        mockTaskInitialization();
+
+        t1.close(true, false);
+        EasyMock.expectLastCall().andThrow(new RuntimeException("any first exception"));
+        t1.close(false, false);
+        EasyMock.expectLastCall().andThrow(new RuntimeException("any second exception"));
+        EasyMock.replay(t1);
+        addAndInitTask();
+        try {
+            assignedTasks.shutdown(true);
+            fail("expected a runtime exception");
+        } catch (final RuntimeException e) {
+            assertThat(e.getMessage(), is("any first exception"));
+        }
+    }
+
+    @Test
+    public void shouldCloseWithoutExceptionOnShutdownIfTaskMigratedException() {
+        mockTaskInitialization();
+
+        t1.close(true, false);
+        EasyMock.expectLastCall().andThrow(new TaskMigratedException());
+        t1.close(false, true);
+        EasyMock.expectLastCall().andThrow(new RuntimeException("any second exception"));
+        EasyMock.replay(t1);
+        addAndInitTask();
+        assignedTasks.shutdown(true);
     }
 
     @Test
@@ -603,7 +635,7 @@ public class AssignedStreamsTasksTest {
             public Set<TaskId> taskIds() {
                 return assignedTasks.suspendedTaskIds();
             }
-            
+
         }.createTaskAndClear();
     }
 

--- a/streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordCollectorTest.java
+++ b/streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordCollectorTest.java
@@ -27,6 +27,7 @@ import org.apache.kafka.common.MetricName;
 import org.apache.kafka.common.Node;
 import org.apache.kafka.common.PartitionInfo;
 import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.errors.ProducerFencedException;
 import org.apache.kafka.common.header.Header;
 import org.apache.kafka.common.header.Headers;
 import org.apache.kafka.common.header.internals.RecordHeader;
@@ -188,6 +189,26 @@ public class RecordCollectorTest {
             }
         });
 
+        collector.send("topic1", "3", "0", null, null, stringSerializer, stringSerializer, streamPartitioner);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Test(expected = RecoverableClientException.class)
+    public void shouldThrowRecoverableExceptionWhenProducerFencedInCallback() {
+        final RecordCollector collector = new RecordCollectorImpl(
+            "test",
+            logContext,
+            new DefaultProductionExceptionHandler(),
+            new Metrics().sensor("skipped-records"));
+        collector.init(new MockProducer(cluster, true, new DefaultPartitioner(), byteArraySerializer, byteArraySerializer) {
+            @Override
+            public synchronized Future<RecordMetadata> send(final ProducerRecord record, final Callback callback) {
+                callback.onCompletion(null, new ProducerFencedException("asdf"));
+                return null;
+            }
+        });
+
+        collector.send("topic1", "3", "0", null, null, stringSerializer, stringSerializer, streamPartitioner);
         collector.send("topic1", "3", "0", null, null, stringSerializer, stringSerializer, streamPartitioner);
     }
 

--- a/streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java
+++ b/streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java
@@ -22,8 +22,11 @@ import org.apache.kafka.clients.consumer.MockConsumer;
 import org.apache.kafka.clients.consumer.OffsetAndMetadata;
 import org.apache.kafka.clients.consumer.OffsetResetStrategy;
 import org.apache.kafka.clients.producer.MockProducer;
+import org.apache.kafka.clients.producer.internals.DefaultPartitioner;
+import org.apache.kafka.common.Cluster;
 import org.apache.kafka.common.KafkaException;
 import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.PartitionInfo;
 import org.apache.kafka.common.TopicPartition;
 import org.apache.kafka.common.errors.AuthorizationException;
 import org.apache.kafka.common.errors.ProducerFencedException;
@@ -51,7 +54,9 @@ import org.apache.kafka.streams.processor.PunctuationType;
 import org.apache.kafka.streams.processor.Punctuator;
 import org.apache.kafka.streams.processor.StateRestoreListener;
 import org.apache.kafka.streams.processor.StateStore;
+import org.apache.kafka.streams.processor.StreamPartitioner;
 import org.apache.kafka.streams.processor.TaskId;
+import org.apache.kafka.streams.processor.WallclockTimestampExtractor;
 import org.apache.kafka.streams.processor.internals.metrics.StreamsMetricsImpl;
 import org.apache.kafka.streams.processor.internals.testutil.LogCaptureAppender;
 import org.apache.kafka.streams.state.internals.OffsetCheckpoint;
@@ -1156,6 +1161,190 @@ public class StreamTaskTest {
         assertFalse(producer.transactionAborted());
         assertFalse(producer.transactionCommitted());
         assertTrue(producer.closed());
+    }
+
+    @Test
+    public void shouldMigrateTaskIfFencedDuringFlush() {
+        final StateStore stateStore = new MockKeyValueStore(storeName, true, true);
+
+        final Map<String, String> storeToChangelogTopic = true ? Collections.singletonMap(storeName, storeName + "-changelog") : Collections.emptyMap();
+        final ProcessorTopology topology1 = new ProcessorTopology(
+            asList(source1, source2),
+            mkMap(mkEntry(topic1, source1), mkEntry(topic2, source2)),
+            Collections.emptyMap(),
+            singletonList(stateStore),
+            Collections.emptyList(),
+            storeToChangelogTopic,
+            Collections.emptySet()
+        );
+
+        task = new StreamTask(
+            taskId00,
+            partitions,
+            topology1,
+            consumer,
+            changelogReader,
+            createConfig(true),
+            streamsMetrics,
+            stateDirectory,
+            null,
+            time,
+            () -> producer = new MockProducer<>(false, bytesSerializer, bytesSerializer)
+        );
+        task.initializeStateStores();
+        task.initializeTopology();
+        producer.fenceProducer();
+
+        try {
+            task.flushState();
+            fail("Expected a TaskMigratedException");
+        } catch (final TaskMigratedException expected) {
+            assertThat(expected.migratedTask(), is(task));
+        }
+    }
+
+    @Test
+    public void shouldMigrateTaskIfFencedDuringProcess() {
+        final StateStore stateStore = new MockKeyValueStore(storeName, true, true);
+
+        final Map<String, String> storeToChangelogTopic = true ? Collections.singletonMap(storeName, storeName + "-changelog") : Collections.emptyMap();
+        final SourceNode<Integer, Integer> sourceNode = new SourceNode<>("test", singletonList(topic1), new WallclockTimestampExtractor(), intDeserializer, intDeserializer);
+        final SinkNode<Integer, Integer> sinkNode = new SinkNode<>("test-sink", new StaticTopicNameExtractor<>("out-topic"), intSerializer, intSerializer, new StreamPartitioner<Integer, Integer>() {
+            @Override
+            public Integer partition(final String topic,
+                                     final Integer key,
+                                     final Integer value,
+                                     final int numPartitions) {
+                return 1;
+            }
+        });
+
+        sourceNode.addChild(sinkNode);
+
+        final ProcessorTopology topology1 = new ProcessorTopology(
+            asList(sourceNode, sinkNode),
+            mkMap(mkEntry(topic1, sourceNode), mkEntry(topic2, sourceNode)),
+            mkMap(mkEntry("out-topic", sinkNode)),
+            singletonList(stateStore),
+            Collections.emptyList(),
+            storeToChangelogTopic,
+            Collections.emptySet()
+        );
+
+        task = new StreamTask(
+            taskId00,
+            partitions,
+            topology1,
+            consumer,
+            changelogReader,
+            createConfig(true),
+            streamsMetrics,
+            stateDirectory,
+            null,
+            time,
+            () -> producer = new MockProducer<byte[], byte[]>(
+                Cluster.empty(),
+                false,
+                new DefaultPartitioner(),
+                bytesSerializer,
+                bytesSerializer
+            ) {
+                @Override
+                public List<PartitionInfo> partitionsFor(final String topic) {
+                    return singletonList(null);
+                }
+            }
+        );
+        task.initializeStateStores();
+        task.initializeTopology();
+        producer.fenceProducer();
+
+        try {
+
+            task.addRecords(partition1, singletonList(getConsumerRecord(partition1, 0)));
+            task.process();
+            fail("Expected a TaskMigratedException");
+        } catch (final TaskMigratedException expected) {
+            assertThat(expected.migratedTask(), is(task));
+        }
+    }
+
+    @Test
+    public void shouldMigrateTaskIfFencedDuringPunctuate() {
+        task = createStatelessTask(createConfig(true));
+
+        final RecordCollectorImpl recordCollector = new RecordCollectorImpl("StreamTask",
+                                                                            new LogContext("StreamTaskTest "), new DefaultProductionExceptionHandler(), new Metrics().sensor("skipped-records"));
+        recordCollector.init(producer);
+
+        task.initializeStateStores();
+        task.initializeTopology();
+        producer.fenceProducer();
+        try {
+            task.punctuate(
+                processorSystemTime,
+                5,
+                PunctuationType.WALL_CLOCK_TIME,
+                timestamp -> recordCollector.send(
+                    "result-topic1",
+                    3,
+                    5,
+                    null,
+                    0,
+                    time.milliseconds(),
+                    new IntegerSerializer(),
+                    new IntegerSerializer()
+                )
+            );
+            fail("Expected a TaskMigratedException");
+        } catch (final TaskMigratedException expected) {
+            assertThat(expected.migratedTask(), is(task));
+        }
+    }
+
+    @Test
+    public void shouldMigrateTaskIfFencedDuringCommit() {
+        final ProcessorTopology topology1 = withSources(
+            asList(source1, source2, processorStreamTime, processorSystemTime),
+            mkMap(mkEntry(topic1, source1), mkEntry(topic2, source2))
+        );
+
+        source1.addChild(processorStreamTime);
+        source2.addChild(processorStreamTime);
+        source1.addChild(processorSystemTime);
+        source2.addChild(processorSystemTime);
+
+        task = new StreamTask(
+            taskId00,
+            partitions,
+            topology1,
+            consumer,
+            changelogReader,
+            createConfig(true),
+            streamsMetrics,
+            stateDirectory,
+            null,
+            time,
+            () -> producer = new MockProducer<>(false, bytesSerializer, bytesSerializer)) {
+            @Override
+            protected void flushState() {
+                // do nothing so that we actually make it to the commit interaction.
+            }
+        };
+
+        final RecordCollectorImpl recordCollector = new RecordCollectorImpl("StreamTask",
+                                                                            new LogContext("StreamTaskTest "), new DefaultProductionExceptionHandler(), new Metrics().sensor("skipped-records"));
+        recordCollector.init(producer);
+
+        task.initializeStateStores();
+        task.initializeTopology();
+        producer.fenceProducer();
+        try {
+            task.commit();
+            fail("Expected a TaskMigratedException");
+        } catch (final TaskMigratedException expected) {
+            assertThat(expected.migratedTask(), is(task));
+        }
     }
 
     @Test

--- a/streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskSuite.java
+++ b/streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskSuite.java
@@ -16,18 +16,23 @@
  */
 package org.apache.kafka.streams.processor.internals;
 
-import org.apache.kafka.common.KafkaException;
+import org.junit.runner.RunWith;
+import org.junit.runners.Suite;
 
 /**
- * Denotes an exception that is recoverable by re-creating the client (ie, the client is no longer in a valid state),
- * as opposed to retriable (the failure was transient, so the same client can be used again later),
- * or fatal (the request was actually invalid, so retrying or recovering would not help)
+ * This suite runs all the tests related to task management. It's intended to simplify feature testing from IDEs.
  *
- * This class also serves the dual purpose of capturing the stack trace as early as possible,
- * at the site of the Producer call, since the exeptions that cause this don't record stack traces.
+ * If desired, it can also be added to a Gradle build task, although this isn't strictly necessary, since all
+ * these tests are already included in the `:streams:test` task.
  */
-public class RecoverableClientException extends KafkaException {
-    public RecoverableClientException(final String message, final Throwable cause) {
-        super(message, cause);
-    }
+@RunWith(Suite.class)
+@Suite.SuiteClasses({
+    AbstractTaskTest.class,
+    StreamTaskTest.class,
+    StandbyTaskTest.class,
+    AssignedStreamsTasksTest.class,
+})
+public class TaskSuite {
 }
+
+

--- a/streams/src/test/java/org/apache/kafka/test/MockKeyValueStore.java
+++ b/streams/src/test/java/org/apache/kafka/test/MockKeyValueStore.java
@@ -18,9 +18,11 @@ package org.apache.kafka.test;
 
 import org.apache.kafka.common.serialization.Deserializer;
 import org.apache.kafka.common.serialization.IntegerDeserializer;
+import org.apache.kafka.common.serialization.StringSerializer;
 import org.apache.kafka.streams.processor.ProcessorContext;
 import org.apache.kafka.streams.processor.StateRestoreCallback;
 import org.apache.kafka.streams.processor.StateStore;
+import org.apache.kafka.streams.processor.internals.RecordCollector;
 import org.apache.kafka.streams.state.KeyValueIterator;
 import org.apache.kafka.streams.state.KeyValueStore;
 
@@ -41,11 +43,22 @@ public class MockKeyValueStore implements KeyValueStore {
     public boolean closed = true;
     public final ArrayList<Integer> keys = new ArrayList<>();
     public final ArrayList<byte[]> values = new ArrayList<>();
+    private final boolean simulateForwardOnFlush;
+    private RecordCollector collector;
 
     public MockKeyValueStore(final String name,
                              final boolean persistent) {
         this.name = name;
         this.persistent = persistent;
+        simulateForwardOnFlush = false;
+    }
+
+    public MockKeyValueStore(final String name,
+                             final boolean persistent,
+                             final boolean simulateForwardOnFlush) {
+        this.name = name;
+        this.persistent = persistent;
+        this.simulateForwardOnFlush = simulateForwardOnFlush;
     }
 
     @Override
@@ -57,12 +70,18 @@ public class MockKeyValueStore implements KeyValueStore {
     public void init(final ProcessorContext context,
                      final StateStore root) {
         context.register(root, stateRestoreCallback);
+        if (simulateForwardOnFlush) {
+            collector = ((RecordCollector.Supplier) context).recordCollector();
+        }
         initialized = true;
         closed = false;
     }
 
     @Override
     public void flush() {
+        if (simulateForwardOnFlush) {
+            collector.send("any", "anykey", "anyvalue", null, 0, 0L, new StringSerializer(), new StringSerializer());
+        }
         instanceLastFlushCount.set(GLOBAL_FLUSH_COUNTER.getAndIncrement());
         flushed = true;
     }

